From 6e1ae0539c4eb62979ae0d8260de2118fb232482 Mon Sep 17 00:00:00 2001
From: David Edmundson <kde@davidedmundson.co.uk>
Date: Mon, 13 Nov 2017 10:03:52 +0000
Subject: Fix crash in KAStatsFavouritesModel

Summary:
KAStatsFavouritesModel doesn't initialise a D-pointer for the lifespan
of the KAStatsFavouritesModel, like the name implies.

It is set in initForClient.

If the activity changes before we have a client ID, there's no point
doing anything and trying to use the clientID will crash.

BUG: 386439

Test Plan: None.

Reviewers: #plasma, hein

Reviewed By: #plasma, hein

Subscribers: anthonyfieroni, plasma-devel

Tags: #plasma

Differential Revision: https://phabricator.kde.org/D8608
---
 applets/kicker/plugin/kastatsfavoritesmodel.cpp | 15 ++++++++++++---
 1 file changed, 12 insertions(+), 3 deletions(-)

diff --git a/applets/kicker/plugin/kastatsfavoritesmodel.cpp b/applets/kicker/plugin/kastatsfavoritesmodel.cpp
index da89c0a..6bdd592 100644
--- a/applets/kicker/plugin/kastatsfavoritesmodel.cpp
+++ b/applets/kicker/plugin/kastatsfavoritesmodel.cpp
@@ -470,8 +470,10 @@ KAStatsFavoritesModel::KAStatsFavoritesModel(QObject *parent)
             this, [&] (const QString &currentActivity) {
                 DEBUG << "Activity just got changed to" << currentActivity;
                 Q_UNUSED(currentActivity);
-                auto clientId = d->m_clientId;
-                initForClient(clientId);
+                if (d) {
+                    auto clientId = d->m_clientId;
+                    initForClient(clientId);
+                }
             });
 }
 
@@ -501,7 +503,7 @@ QString KAStatsFavoritesModel::description() const
 
 bool KAStatsFavoritesModel::trigger(int row, const QString &actionId, const QVariant &argument)
 {
-    return d->trigger(row, actionId, argument);
+    return d && d->trigger(row, actionId, argument);
 }
 
 bool KAStatsFavoritesModel::enabled() const
@@ -547,6 +549,7 @@ bool KAStatsFavoritesModel::isFavorite(const QString &id) const
 
 void KAStatsFavoritesModel::portOldFavorites(const QStringList &ids)
 {
+    if (!d) return;
     DEBUG << "portOldFavorites" << ids;
 
     const auto activityId = ":global";
@@ -617,6 +620,8 @@ void KAStatsFavoritesModel::addFavoriteTo(const QString &id, const Activity &act
 
 void KAStatsFavoritesModel::removeFavoriteFrom(const QString &id, const Activity &activity)
 {
+    if (!d || id.isEmpty()) return;
+
     const auto url = d->normalizedId(id).value();
 
     Q_ASSERT(!activity.values.isEmpty());
@@ -631,6 +636,8 @@ void KAStatsFavoritesModel::removeFavoriteFrom(const QString &id, const Activity
 
 void KAStatsFavoritesModel::setFavoriteOn(const QString &id, const QString &activityId)
 {
+    if (!d || id.isEmpty()) return;
+
     const auto url = d->normalizedId(id).value();
 
     DEBUG << "setFavoriteOn" << id << activityId << url << " (actual)";
@@ -650,6 +657,8 @@ void KAStatsFavoritesModel::setFavoriteOn(const QString &id, const QString &acti
 
 void KAStatsFavoritesModel::moveRow(int from, int to)
 {
+    if (!d) return;
+
     d->move(from, to);
 }
 
-- 
cgit v0.11.2

